build-system/release-workflows/cut-nightly.js

```diff
@@ -72,47 +62,140 @@ async function cutNightlyBranch() {
       cyan(sha),
       'have completed successfully. Cutting a new nightly at this commit...'
     );
-    const response = await octokit.rest.git.updateRef({
-      ...params,
-      ref: 'heads/nightly',
-      sha,
-    });
-
-    // Casting to Number because the return type in Octokit is incorrectly
-    // annotated to only ever return 200.
-    switch (Number(response.status)) {
-      case 201:
-        log(
-          'A new',
-          cyan('nightly'),
-          'branch was successfully cut at commit',
-          cyan(sha)
-        );
-        break;
-      case 200:
-        log(
-          'The',
-          cyan('nightly'),
-          'branch is already at the latest',
-          green('green'),
-          'commit',
-          cyan(sha)
-        );
-        break;
-      default:
-        log(
-          red(
-            'An uncaught status was returned while attempting to fast-forward the'
-          ),
-          cyan('nightly'),
-          red('branch to commit'),
-          cyan(sha)
-        );
-        log('See full response:', response);
-    }
 
-    break;
+    return sha;
+  }
+}
+
+/**
+ * Fast forward nightly branch to given sha
+ * @param {Octokit} octokit
+ * @param {string} sha
+ * @return {Promise<void>}
+ */
+async function updateBranch(octokit, sha) {
+  const response = await octokit.rest.git.updateRef({
+    ...params,
+    ref: 'heads/nightly',
+    sha,
+  });
+
+  // Casting to Number because the return type in Octokit is incorrectly
+  // annotated to only ever return 200.
+  switch (Number(response.status)) {
+    case 201:
+      log(
+        'A new',
+        cyan('nightly'),
+        'branch was successfully cut at commit',
+        cyan(sha)
+      );
+      break;
+    case 200:
+      log(
+        'The',
+        cyan('nightly'),
+        'branch is already at the latest',
+        green('green'),
+        'commit',
+        cyan(sha)
+      );
+      break;
+    default:
+      log(
+        red(
+          'An uncaught status was returned while attempting to fast-forward the'
+        ),
+        cyan('nightly'),
+        red('branch to commit'),
+        cyan(sha)
+      );
+      log('See full response:', response);
   }
 }
 
+/**
+ * Create GitHub tag
+ * @param {any} octokit
+ * @param {string} sha
+ * @return {Promise<void>}
+ */
+async function createTag(octokit, sha) {
+  // get commit time using octokit instead of git
+  const commit = await octokit.rest.git.getCommit({...params, sha});
+
+  // generate an AMP version according to `build-system/compile/internal-version.js`
+  // NOTE: this does NOT account for cherry-picked commits
+  const date = moment(commit.committer.date).format('YYMMDDHHmmss');
+  const ampVersion = date.slice(0, -2) + '000';
```

Comment:
The problem with this is that we now have to maintain (part of) this code in two places. We already have the git tree checked out, so we can just do `git checkout ${sha}` and... I dunno, `amp version` should be a task by now?

Reply:
It's pretty simple: #36109

Reply:
> Agreed not ideal. If we `git checkout ${sha}`, isn't there the possibility that `cut-nightly.js` might not be the same?

Oh yeah I see now what you mean. Once the file is loaded into `node`, changing the source wouldn't change the running code - so _that_ part is safe

Reply:
:/ #36109 is just a wrapper? Without `git checkout` it doesn't work for this PR's use case

Reply:
> :/ #36109 is just a wrapper? Without `git checkout` it doesn't work for this PR's use case

Yes 😅
Try something like `git checkout ${sha} && export AMP_VERSION=node build-system/common/version.js && git checkout -`?

Reply:
Oh oh oh or or... if we add an optional parameter to[`gitCherryMain`](https://github.com/ampproject/amphtml/blob/main/build-system/common/git.js#L169-L172) so that it compares an arbitrary `sha` with `main`, and modify [`getVersion`](https://github.com/ampproject/amphtml/blob/main/build-system/compile/internal-version.js#L65) to pass along the same optional parameter, that could do the trick!

Reply:
Yaaa I was going down the optional parameter to `getVersion` path. I can try it and add it to this PR

Reply:
https://github.com/ampproject/amphtml/pull/36095/commits/2d90027c9e6b7c0520ddd4c109309defdaaf1b3a should work. `git cherry` defaults to `HEAD` but it can also be a ref

Reply:
🥳